src/Http/Http/src/Features/QueryFeature.cs

```diff
@@ -100,5 +103,170 @@ public IQueryCollection Query
                 }
             }
         }
+
+        /// <summary>
+        /// Parse a query string into its component key and value parts.
+        /// </summary>
+        /// <param name="queryString">The raw query string value, with or without the leading '?'.</param>
+        /// <returns>A collection of parsed keys and values, null if there are no entries.</returns>
+        internal static AdaptiveCapacityDictionary<string, StringValues>? ParseNullableQueryInternal(string? queryString)
+        {
+            var accumulator = new KvpAccumulator();
+
+            if (string.IsNullOrEmpty(queryString) || queryString == "?")
+            {
+                return null;
+            }
+
+            int scanIndex = 0;
+            if (queryString[0] == '?')
+            {
+                scanIndex = 1;
+            }
+
+            int textLength = queryString.Length;
+            int equalIndex = queryString.IndexOf('=');
+            if (equalIndex == -1)
+            {
+                equalIndex = textLength;
+            }
+            while (scanIndex < textLength)
+            {
+                int delimiterIndex = queryString.IndexOf('&', scanIndex);
+                if (delimiterIndex == -1)
```

Comment:
```suggestion
                if (delimiterIndex < 0)
```
This is really a micro-optimization, as the this produces `test reg, reg; jge label` which the CPU will fuse, so it's actually only one instruction to be executed, as opposed to `cmp reg, const; jne label` which is two instructions.

But you don't need to take it here, more or less FYI.